lib/c-api/src/wasm_c_api/wasi/mod.rs

```diff
@@ -349,24 +345,137 @@ pub unsafe extern "C" fn wasi_get_wasi_version(module: &wasm_module_t) -> wasi_v
         .unwrap_or(wasi_version_t::INVALID_VERSION)
 }
 
-/// Takes ownership of `wasi_env_t`.
+/// Non-standard type wrapping `wasm_extern_t` with the addition of
+/// two `wasm_name_t` respectively for the module name and the name of
+/// the extern (very likely to be an import). This non-standard type
+/// is used by the non-standard `wasi_get_unordered_imports` function.
+///
+/// The `module`, `name` and `extern` fields are all owned by this type.
+#[allow(non_camel_case_types)]
+#[derive(Clone)]
+pub struct wasm_named_extern_t {
+    module: Box<wasm_name_t>,
+    name: Box<wasm_name_t>,
+    r#extern: Box<wasm_extern_t>,
+}
+
+wasm_declare_boxed_vec!(named_extern);
+
+/// Non-standard function to get the module name of a
+/// `wasm_named_extern_t`.
+///
+/// The returned value isn't owned by the caller.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_module(
+    named_extern: Option<&wasm_named_extern_t>,
+) -> Option<&wasm_name_t> {
+    Some(named_extern?.module.as_ref())
+}
+
+/// Non-standard function to get the name of a `wasm_named_extern_t`.
+///
+/// The returned value isn't owned by the caller.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_name(
+    named_extern: Option<&wasm_named_extern_t>,
+) -> Option<&wasm_name_t> {
+    Some(named_extern?.name.as_ref())
+}
+
+/// Non-standard function to get the wrapped extern of a
+/// `wasm_named_extern_t`.
+///
+/// The returned value isn't owned by the caller.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_unwrap(
+    named_extern: Option<&wasm_named_extern_t>,
+) -> Option<&wasm_extern_t> {
+    Some(named_extern?.r#extern.as_ref())
+}
+
+/// Non-standard function to get the imports needed for the WASI
+/// implementation with no particular order. Each import has its
+/// associated module name and name, so that it can be re-order later
+/// based on the `wasm_module_t` requirements.
+///
+/// This function takes ownership of `wasm_env_t`.
+#[no_mangle]
+pub unsafe extern "C" fn wasi_get_unordered_imports(
+    store: Option<&wasm_store_t>,
+    module: Option<&wasm_module_t>,
+    wasi_env: Option<Box<wasi_env_t>>,
+    imports: &mut wasm_named_extern_vec_t,
+) -> bool {
+    wasi_get_unordered_imports_inner(store, module, wasi_env, imports).is_some()
+}
+
+unsafe fn wasi_get_unordered_imports_inner(
```

Comment:
Unclear if this function actually needs to be unsafe, same with any function not taking a pointer, really. We can always audit this later though

Reply:
True! I'll remove the `unsafe` keyword.